### PR TITLE
[GSOC] Enable OCL for AKAZE

### PR DESCRIPTION
**Pull request by hrnr**
This re-enables OCL for AKAZE.

* ocl path is only used when image is passed as UMat
* CPU path does not use any UMats
* added new ocl test that checks results are the same CPU version

cc: @bmagyar @alalek

### CODE REVIEW

**Comment by hrnr** [comment]
Hi @alalek , could you please provide some advice how I should integrate OCL in AKAZE?

**Comment by alalek** [comment]
Please follow these statements:
- don't enable OpenCL code path silently for `cv::Mat`. Enable OpenCL code path if user passes cv::UMat images explicitly.
- using cv::UMat internally for other OpenCV functions forces OpenCL code paths for them. Don't do this unconditionally.
- `cv::UMat::getMat()` calls should be used for local processing only (not more than function code scope). Minimize lifetime of these derived objects as mush as possible. If you want long-life data download - use `copyTo()`.
- unfortunately, current cv::UMat implementation/design is not very "universal". Data is forced be stored in OpenCL buffers (there is no "on-demand" behavior).
- use passed InputArray / OutputArray as mush as possible - they are "universal pointers". Don't call getMat()/getUMat() without a strong reason (to just pass these buffers to next functions).
- add separate tests for OpenCL (T-API) code paths (use `test/opencl` / `perf/opencl` sub-directories).

**Comment by bmagyar** [comment]
Thanks for the heads up @alalek !

**Comment by hrnr** [comment]
thanks. Then it's probably better to split OCL/CPU paths completely.                                                                                                                                                                                  ‎

**Comment by hrnr** [comment]
@alalek What do you think about current approach?

**Comment by alalek** [comment]
@hrnr Looks better. Could you add "OCL" (UMat) test for AKAZE?

File: modules/features2d/src/kaze/AKAZEFeatures.cpp [hunk 1]
```diff
@@ -455,42 +481,84 @@ void AKAZEFeatures::Create_Nonlinear_Scale_Space(InputArray img)
...
+  }
+}
+
+/**
+ * @brief This method creates the nonlinear scale space for a given image
+ * @param image Input image for which the nonlinear scale space needs to be created
+ */
+void AKAZEFeatures::Create_Nonlinear_Scale_Space(InputArray image)
+{
+  if (image.isUMat()) {
```
Comment (alalek) [inline modules/features2d/src/kaze/AKAZEFeatures.cpp]: `&& cv::ocl::useOpenCL()`

**Comment by hrnr** [comment]
Thanks for review.

I've added test checking results are the same between CPU and GPU. For me this is good to go in, I don't plan to add anything more.

**Comment by hrnr** [comment]
there seems to be failure on buildbot. otherwise the build and tests are OK.

**Comment by alalek** [comment]
Don't worry - there are sporadic hang failures of Mac builders. This build is restarted.

**Comment by alalek** [comment]
Looks good to me. Thank you! :+1:

**Comment by bjquinn** [comment]
If I understand correctly, it looks like this merge missed the deadline for 3.3.0.  If I want to use AKAZE with OpenCL, I would either need to build OpenCV from source or wait until the next release? (probably ~6 months if history is a guide)

Thanks!!

**Comment by bjquinn** [comment]
For what it's worth, I cloned opencv-python, updated opencv to master, and used Appveyor to build it for Windows.  GPU acceleration ended up being slower for our purposes than CPU, so I don't think we'll use it at the moment, but for what it's worth, anyone interested can see the below.

https://github.com/bjquinn/opencv-python

https://ci.appveyor.com/api/buildjobs/wrdwimqmaidxu4qi/artifacts/dist%2Fopencv_python-3.3.0%2Bccba2c6-cp36-cp36m-win_amd64.whl

**Comment by hrnr** [comment]
Thanks for sharing the results. The OCL path is expected to win over CPU only on large images, so your results makes sense. After the changes in CPU path the threshold is quite high (but that of cause depends also on your CPU vs GPU performace).

**Comment by bjquinn** [comment]
@hrnr our images are relatively large -- letter size pages at 200 DPI, but of course being documents, they're mostly empty space.  At any rate, it seems that for our purpose, adding CPU cores will be easier/cheaper to scale than GPUs given the performance characteristics or our data.  But the improved CPU performance of AKAZE is much appreciated.

**Pull request merged**